collectors/python.d.plugin/uwsgi/uwsgi.chart.py

```diff
@@ -0,0 +1,169 @@
+        self.url = self.configuration.get('host', 'localhost')
+        self.port = self.configuration.get('port', 1717)
+        self.order = ORDER
+        self.definitions = deepcopy(CHARTS)
+
+        self.last_result = {}
+        self.workers = []
+
+    def read_data(self):
+        """
+        Read data from socket and parse as JSON.
+        :return: (dict) stats
+        """
+        raw_data = self._get_raw_data()
+        if not raw_data:
+            return None
+        try:
+            return json.loads(raw_data)
+        except ValueError as err:
+            self.error(err)
+            return None
+
+    def check(self):
+        """
+        Parse configuration, check if redis is available, and dynamically create chart lines data
+        :return: boolean
+        """
+        self._parse_config()
+
+        # Clear dynamic dimensions, these are added during `_get_data()` to allow adding workers at run-time
+        for chart in ('requests', 'tx', 'avg_rt', 'memory_rss', 'memory_vsz'):
+            self.definitions[chart]['lines'] = []
+
+        stats = self.read_data()
+        return bool(stats)
+
+    def add_worker_dimensions(self, key):
+        """
+        Helper to add dimensions for a worker.
+        :param key: (int or str) worker identifier
+        :return:
+        """
+        for chart in ('requests', 'tx', 'avg_rt', 'memory_rss', 'memory_vsz'):
+
+            for line in CHARTS[chart]['lines']:
+                dimension_id = '%s_%s' % (line[0], key)
+                dimension_name = str(key)
+
+                dimension = [dimension_id, dimension_name] + line[2:]
+                self.charts[chart].add_dimension(dimension)
+
+    @staticmethod
+    def _check_raw_data(data):
+        # The server will close the connection when it's done sending
+        # data, so just keep looping until that happens.
+        return False
+
+    def _get_data(self):
+        """
+        Read data from socket
+        :return: dict
+        """
+        stats = self.read_data()
+        if not stats:
+            return None
+
+        result = {
+            'exceptions': 0,
+            'harakiri_count': 0,
+            'respawn_count': 0,
+        }
+
+        for worker in stats['workers']:
+            key = worker['pid']
+
+            # Add dimensions for new workers
+            if key not in self.workers:
+                self.add_worker_dimensions(key)
+                self.workers.append(key)
+
+            result['requests_%s' % key] = worker['requests']
+            result['tx_%s' % key] = worker['tx']
+            result['avg_rt_%s' % key] = worker['avg_rt']
+
+            # avg_rt is not reset by uwsgi, so reset here
+            if self.last_result.get('requests_%s' % key) == worker['requests']:
+                result['avg_rt_%s' % key] = 0
+
+            result['memory_rss_%s' % key] = worker['rss']
+            result['memory_vsz_%s' % key] = worker['vsz']
```

Comment:
please use `format` instead of `%`

Reply:
personally I think
```result['requests_%s' % key] = worker['rss']```
is more readable than
```result['requests_{}'.format(key)] = worker['rss']```
but will change if that is what you prefer

Reply:
https://docs.python.org/release/3.1.5/library/stdtypes.html#old-string-formatting-operations

Reply:
or you can use some helper func to make it less verbose

like

```python
def fmt(f, s):
    return '{0}_{1}'.format(f, s)

result['requests_{}'.format(key)] = worker['rss']
#  vs
result[fmt('requests', key)] = worker['rss']
```

Reply:
> https://docs.python.org/release/3.1.5/library/stdtypes.html#old-string-formatting-operations
This dates from 2012.
`printf`-style String Formatting was never deprecated and will not be removed afaik.

https://docs.python.org/3/library/stdtypes.html#old-string-formatting
https://stackoverflow.com/questions/13451989/pythons-many-ways-of-string-formatting-are-the-older-ones-going-to-be-deprec